src/core/raster/qgssinglebandpseudocolorrenderer.cpp

```diff
@@ -84,6 +88,193 @@ void QgsSingleBandPseudoColorRenderer::setShader( QgsRasterShader* shader )
   mShader = shader;
 }
 
+void QgsSingleBandPseudoColorRenderer::createShader( QgsColorRamp* colorRamp, QgsColorRampShader::ColorRamp_TYPE colorRampType, bool clip )
```

Comment:
I still have the impression that this method would be better suited to QgsColorRampShader class as a static factory method:

```
QgsColorRampShader* create( colorRamp, type, clip, mode, numClasses, min, max, input )
```
It would return a new instance of a shader or NULL if it didn't work out for some reason. The code in QgsSingleBandPseudoColorRenderer::setShaderFromColorRamp( colorRamp, type, clip ) would just use that function with its classification parameters and if successful, replace the current shader. That approach should also allow make it easier to unit test just the shader creation.

The problems I see with the current implementation:
- in some cases it does premature "return" with no change to shader, in other cases it creates new shader with no entries
- client code can't easily find out whether shader creation was successful or not
- unit testing of the shader creation is more difficult as one needs to work also with the renderer class